gossipd/gossip.c

```diff
@@ -1594,37 +1599,32 @@ static struct io_plan *connection_out(struct io_conn *conn,
 				   handshake_out_success, reach);
 }
 
-static void try_connect(struct reaching *reach);
-
 static void connect_failed(struct io_conn *conn, struct reaching *reach)
 {
-	u32 diff = time_now().ts.tv_sec - reach->first_attempt;
-	reach->attempts++;
-
-	if (!important_peerid_map_get(&reach->daemon->important_peerids, &reach->id)
-	    && reach->attempts >= 10) {
-		status_info("Failed to connect after %d attempts, giving up "
-			    "after %d seconds",
-			    reach->attempts, diff);
-		daemon_conn_send(
-		    &reach->daemon->master,
-		    take(towire_gossip_peer_connection_failed(
-			NULL, &reach->id, diff, reach->attempts, false)));
-		tal_free(reach);
-	} else {
-		unsigned int secs;
-
-		/* Exponential backoff, then every 5 minutes */
-		if (reach->attempts < 9)
-			secs = 1 << reach->attempts;
-		else
-			secs = 300;
-		status_trace("Failed connected out for %s, will try again in %u seconds",
-			     type_to_string(tmpctx, struct pubkey, &reach->id),
-			     secs);
-		new_reltimer(&reach->daemon->timers, reach, time_from_sec(secs),
-			     try_connect, reach);
+	u8 *msg;
+	struct important_peerid *imp;
+
+	/* Tell any connect commands what happened. */
+	msg = towire_gossipctl_connect_to_peer_result(reach, &reach->id,
+						      false, strerror(errno));
+	for (size_t i = 0; i < reach->num_master_responses; i++)
```

Comment:
This fan-out in a subdaemon is rather strange, shouldn't we be doing it in the master? At least that's what we've been doing for invoices. IMHO the subdaemon should just report all raw events and the master should map them to commands. No need to pull in the commands concept into the subdaemon.

Reply:
The master can ask to connect to the same peer multiple times, we need to respond to all of them.  That seems the simpler API model.

We could enforce that the master can only ask for one at a time, though?

Reply:
I will address this, but note that we have to handle the case where we are reconnecting (because important peer) and we get a connect command for the same peer.  So the counter becomes a boolean.

Reply:
We could put the `connect` commands in a list on `struct lightningd`, then send a single message to the masterd, and then the handler of that message would iterate over the entire list and trigger each of the matching `connect` commands.